compiler/env/newmemory/CustomAllocator.hpp

```diff
@@ -0,0 +1,87 @@
+/*******************************************************************************
+ * Copyright (c) 2019, 2019 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at http://eclipse.org/legal/epl-2.0
+ * or the Apache License, Version 2.0 which accompanies this distribution
+ * and is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License, v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception [1] and GNU General Public
+ * License, version 2 with the OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#ifndef TR_CUSTOMALLOCATOR_INCL
+#define TR_CUSTOMALLOCATOR_INCL
+
+#pragma once
+
+#include <stddef.h>  // for size_t
+#include <cstdlib>   // for free, malloc
+#include <new>       // for bad_alloc, nothrow, nothrow_t
+#include "infra/Assert.hpp" //  for TR_ASSERT_FATAL
+#include "env/mem/RawAllocator.hpp"
+
+namespace TR {
+
+// CustomAllocator is a more flexible implementation of TR::RawAllocator
+// but at the performance cost of an extra pointer indirection for every
+// API call. Upon construction, the functions to call to implement
+// allocate() (AllocatFunction), deallocate (DeallocateFunction), and
+// optionally protect() (ProtectFunction) must be specified.
+
+class CustomAllocator :  public TR::RawAllocator
```

Comment:
I'm not sure I understand the purpose of this class. If a user wanted to use custom memory management functions, would it not be easier for them to just created a new sub-class of `TR::RawAllocator`?

Reply:
they could do that, yes. OpenJ9 has to call malloc and free via function pointers anyway for its "debugger" scenario, so I generalized it to allow any function pointers to be used. I felt that was less specialized.

Reply:
Well, OMR doesn't *need* a CustomRawAllocator, but I guess I thought it might be easier for someone to show up and create a CustomRawAllocator rather than write up a new class, figure out where to add new files to build for which build system they choose to use.

Another motivating example: if the JIT is compiled into a dynamic library, then you'll only have pointers to the allocate/free functions anyway (assuming they're not in the JIT's library, which is the case in OpenJ9), so why force everyone to create a new class to enable dynamic linking?